### ISSUE TITLE
Segfault when joining two cuboids

### ISSUE THREAD

**Issue by florianwechsung** [nschloe/pygalmesh#15]
I'm trying to join two cuboids to create a backwards facing step.
Unfortunately I get a segfault when I run the code below.
This is on macOS 10.13.4 with Python 3.6.5, cgal 4.12 from brew and pygalmesh v0.2.4.
```python
import pygalmesh

c1 = pygalmesh.Cuboid([0., 1., 0.], [1.0, 2.0, 1.0])
c2 = pygalmesh.Cuboid([1., 0., 0.], [10., 2., 1.])
u = pygalmesh.Union([c1, c2])
pygalmesh.generate_mesh(
        u, 'out.mesh', cell_size=0.1, edge_size=0.1
        )
```

I had to apply this patch to get it to install, but that shouldn't be relevant.
```diff
diff --git a/setup.py b/setup.py
index 73bbac0..7fadc74 100644
--- a/setup.py
+++ b/setup.py
@@ -42,13 +42,14 @@ ext_modules = [Extension(
         ],
     language='c++',
     include_dirs=[
+        '/usr/local/Cellar/eigen/3.3.4/include/eigen3/',
         '/usr/include/eigen3/',
         # Path to pybind11 headers
         get_pybind_include(),
         get_pybind_include(user=True)
         ],
     libraries=['CGAL', 'gmp', 'mpfr'],
-    # extra_compile_args=['-std=c++11']
+    extra_compile_args=['-std=c++11']
     )]

 setup(
```

**Comment by nschloe** [comment]
Thanks for the report. I've tried to reproduce this, but the script just ran forever, filling the memory until it eventually choked. Somethings not right with `Union`. I'll check it out sometime later.

**Comment by nschloe** [comment]
I've improved some things in domain specifications and released a new version, but the above mesh will still fail. I found that all cuboid unions fail where the two cuboids share part of at least one side exactly. For example, this
```python
c1 = pygalmesh.Cuboid([0., 0., 0.], [1.0, 1.0, 1.0])
c2 = pygalmesh.Cuboid([-0.5, -0.5, -0.5], [0.9, 0.9, 1.0])
```
will hang/segfault as well. I'm not sure why this happens yet, perhaps is has to do with the feature edges. As a work-around, add a miniscule offset; this works:
```python
c1 = pygalmesh.Cuboid([0., 0., 0.], [1.0, 1.0, 1.0])
c2 = pygalmesh.Cuboid([-0.5, -0.5, -0.5], [0.9, 0.9, 1.0 - 1.0e-15])
```

**Comment by florianwechsung** [comment]
Thanks for that! 
Unfortunately when adding such a small offset, this seems to confuse the mesh generator, as I see this when I zoom into (0.0, 0.9, 1.0):

<img width="1196" alt="image" src="https://user-images.githubusercontent.com/5160278/40281564-6258bd36-5c5b-11e8-8971-5104d0bedbb9.png">

If I zoom out further, I see a few more issues:

<img width="1039" alt="image" src="https://user-images.githubusercontent.com/5160278/40281569-7803d170-5c5b-11e8-82fc-262e269357f8.png">

**Comment by nschloe** [comment]
Yup, feature edges aren't preserved by default. Right now, _all_ primitive domains are implemented in terms of level-sets, even polygonal domains like cuboids. As a consequence, feature edges are _not_ preserved automatically. There are native polyhedral domains in CGAL, perhaps that can help out. Not sure...
